contracts/instance/InstanceAccessManager.sol

```diff
@@ -1,292 +1,526 @@
 import {TimestampLib} from "../types/Timestamp.sol";
+import {NftId} from "../types/NftId.sol";
+
+import {AccessManagerUpgradeableInitializeable} from "./AccessManagerUpgradeableInitializeable.sol";
+
 import {IAccess} from "./module/IAccess.sol";
+import {IRegistry} from "../registry/IRegistry.sol";
 
 contract InstanceAccessManager is
     AccessManagedUpgradeable
 {
+    event LogRoleCreation(RoleId roleId, ShortString name, IAccess.Type rtype);
+    event LogTargetCreation(address target, ShortString name, IAccess.Type ttype, bool isLocked);
+
     using RoleIdLib for RoleId;
 
     string public constant ADMIN_ROLE_NAME = "AdminRole";
     string public constant PUBLIC_ROLE_NAME = "PublicRole";
 
-    uint64 public constant CUSTOM_ROLE_ID_MIN = 10000;
+    uint64 public constant CUSTOM_ROLE_ID_MIN = 10000; // MUST be even
     uint32 public constant EXECUTION_DELAY = 0;
 
     // role specific state
-    mapping(RoleId roleId => IAccess.RoleInfo info) internal _role;
+    mapping(RoleId roleId => IAccess.RoleInfo info) internal _roleInfo;
     mapping(RoleId roleId => EnumerableSet.AddressSet roleMembers) internal _roleMembers; 
-    mapping(ShortString name => RoleId roleId) internal _roleForName;
-    RoleId [] internal _roles;
+    mapping(ShortString name => RoleId roleId) internal _roleIdForName;
+    RoleId [] internal _roleIds;
+    uint64 _idNext;
 
     // target specific state
-    mapping(address target => IAccess.TargetInfo info) internal _target;
-    mapping(ShortString name => address target) internal _targetForName;
+    mapping(address target => IAccess.TargetInfo info) internal _targetInfo;
+    mapping(ShortString name => address target) internal _targetAddressForName;
     address [] internal _targets;
 
-    AccessManager internal _accessManager;
+    AccessManagerUpgradeableInitializeable internal _accessManager;
+    IRegistry internal _registry;
 
-    function initialize(address initialAdmin) external initializer
+    modifier restrictedToRoleAdmin(RoleId roleId) {
+        RoleId admin = getRoleAdmin(roleId);
+        (bool inRole, uint32 executionDelay) = _accessManager.hasRole(admin.toInt(), _msgSender());
+        assert(executionDelay == 0); // to be sure no delayed execution functionality is used
+        if (!inRole) {
+            revert IAccess.ErrorIAccessCallerIsNotRoleAdmin(_msgSender(), roleId);
+        }
+        _;
+    }
+
+    function initialize(address ozAccessManager, address registry) external initializer 
     {
-        // if size of the contract gets too large, this can be externalized which will reduce the contract size considerably
-        _accessManager = new AccessManager(address(this));
-        // this service required admin rights to access manager to be able to grant/revoke roles
-        _accessManager.grantRole(_accessManager.ADMIN_ROLE(), initialAdmin, 0);
+        require(ozAccessManager != address(0));
```

Comment:
please use custom error instead of `require`. Also never use require without an explicit error message. this is a general rule for our code.

Reply:
deleted for now